### PR TITLE
Default extra_plugin_paths if necessary

### PR DESCRIPTION
**Pull request by jperon**
Some changes working on #6768:

- add a menu option to set extra_plugin_paths;
- add a default extra plugin paths in data_dir when default plugin path
  is read-only.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6849)

### CODE REVIEW

**Comment by jperon** [comment]
Just wondering for the menu option: is it better in `"device"` (like screenshot folder selection), in `"more_tools"` or in `"developer_options"` ?

**Comment by Frenzie** [comment]
Sounds like developer options to me. Just adding a default `/sdcard/koreader/plugins` type folder there seems like it should pretty much do the trick.

**Comment by jperon** [comment]
`/sdcard/koreader/plugins` would be specific to Android, and wouldn't cover the case of Appimage, for example. With `getDataDir()`, it covers all cases.

**Comment by Frenzie** [comment]
I know. I mean that in contrast to all the lines about user selection.

File: frontend/pluginloader.lua [hunk 1]
```diff
@@ -49,6 +58,11 @@ function PluginLoader:loadPlugins()
         else
             logger.err("extra_plugin_paths config only accepts string or table value")
         end
+    else
+        local datadir = DataStorage:getDataDir()
+        if datadir ~= "." then
+            G_reader_settings:saveSetting("extra_plugin_paths", { datadir .. "/plugins/" })
+        end
```
Comment (Frenzie) [inline frontend/pluginloader.lua:65]: I.e., these lines good, the rest looks like feature creep.

Comment (Frenzie) [reply frontend/pluginloader.lua:65]: Therefore *if you must*, put it in developer options.

Comment (jperon) [reply frontend/pluginloader.lua:65]: OK, I see.

File: frontend/pluginloader.lua [hunk 2]
```diff
@@ -49,6 +49,11 @@ function PluginLoader:loadPlugins()
         else
             logger.err("extra_plugin_paths config only accepts string or table value")
         end
+    else
+        local datadir = require("datastorage"):getDataDir()
```
Comment (Frenzie) [inline frontend/pluginloader.lua:53]: Probably best to keep it data_dir just like there.

**Review by Frenzie — APPROVED** [review]
lgtm

File: frontend/pluginloader.lua (hunk 2 above)
Comment (yparitcher) [reply frontend/pluginloader.lua]: you can also move this after the `if` so it is not called on other platforms, and then change it to just a `elseif`

**Review by poire-z — APPROVED** [review]

**Review by pazos — COMMENTED** [review]
A minor nit:

Create the `plugins` dir if `data_dir ~= "."`

I think it can be done at the end of [`initDataDir()`](https://github.com/koreader/koreader/blob/master/datastorage.lua#L53)

```lua
if DataStorage:getDataDir() ~= "." then
    local sub_data_dir = string.format("%s/%s", DataStorage:getDataDir(), "plugins")
    if lfs.attributes(sub_data_dir, "mode") ~= "directory" then
        lfs.mkdir(sub_data_dir)
    end
end
```

**Comment by jperon** [comment]
@pazos I just added `"plugins"` to `sub_data_dirs`: if `data_dir` is `.`, `plugins/` already exists, and it will be detected by `if lfs.attributes(sub_data_dir, "mode") ~= "directory"`

File: datastorage.lua [hunk 3]
```diff
@@ -54,7 +54,7 @@ local function initDataDir()
     local sub_data_dirs = {
         "cache", "clipboard",
         "data", "data/dict", "data/tessdata",
-        "history", "ota",
+        "history", "ota", "plugins",
```
Comment (Frenzie) [inline datastorage.lua:57]: If you're changing this line anyway, I don't think ota is relevant anymore on platforms where this function is relevant. Unless I'm mistaken @pazos?
```suggestion
        "history", "plugins",
```

**Comment by Frenzie** [comment]
> initDataDir()

Yup, good catch. I added that because it was a bit confusing to users on platforms like Android since there was no self-documenting directory structure already in place.

File: datastorage.lua (hunk 3 above)
Comment (pazos) [reply datastorage.lua]: appImage?

Comment (Frenzie) [reply datastorage.lua]: While I have considered looking into https://github.com/AppImage/AppImageUpdate (which iirc uses zsync?), I don't think that folder would be relevant to it. But right now it simply opens the link to the latest AppImage.

**Review by pazos — APPROVED** [review]

**Comment by pazos** [comment]
@jperon @Frenzie 

It seems that "ota" folder was needed, at least on android. Here's what happens on a new install if not present:

 `11-07 21:27:13.492 17748 17774 W KOReader:  cannot find update file /storage/emulated/0/koreader/ota/koreader-android-latest-nightly: No such file or directory`

and a infomessage "update information not available on xxx channel".

Not a very serious issue since legacy ota dirs are preserved, but we should restore that behaviour to not break OTA updates for new users.

File: frontend/pluginloader.lua [hunk 4]
```diff
@@ -49,6 +49,11 @@ function PluginLoader:loadPlugins()
         else
             logger.err("extra_plugin_paths config only accepts string or table value")
         end
+    else
+        local data_dir = require("datastorage"):getDataDir()
+        if data_dir ~= "." then
+            G_reader_settings:saveSetting("extra_plugin_paths", { data_dir .. "/plugins/" })
+        end
```
Comment (Hzj-jie) [inline frontend/pluginloader.lua:56]: Did you forget to add the new extra path to lookup_path_list?

**Pull request merged**